frontend/ui/widget/htmlboxwidget.lua

```diff
@@ -14,15 +14,197 @@ local logger = require("logger")
+    end
+end
+
+local function getSelectedText(lines, start_pos, end_pos)
+    local start_line_index, start_word_index = getWordIndices(lines, start_pos)
+    local end_line_index, end_word_index = getWordIndices(lines, end_pos)
+    if start_line_index == nil or end_line_index == nil then
+        return nil, nil
+    elseif start_line_index > end_line_index then
+        start_line_index, end_line_index = end_line_index, start_line_index
+        start_word_index, end_word_index = end_word_index, start_word_index
+    elseif start_line_index == end_line_index and start_word_index > end_word_index then
+        start_word_index, end_word_index = end_word_index, start_word_index
+    end
+
+    local found_start = false
+    local words = {}
+    local rects = {}
+
+    for line_index = start_line_index, end_line_index do
+        local line = lines[line_index]
+        local line_last_rect = nil
+        local line_text_direction = getLineTextDirection(line)
+
+        for word_index, word in ipairs(line) do
+            if type(word) == 'table' then
+                if line_index == start_line_index and word_index == start_word_index then
+                    found_start = true
+                end
+
+                if found_start then
+                    table.insert(words, word.word)
+
+                    -- do not try to join word rects in mixed direction lines
+                    if line_last_rect == nil or line_text_direction == 0 then
+                        local rect = Geom:new{
+                            x = word.x0,
+                            y = line.y0,
+                            w = word.x1 - word.x0,
+                            h = line.y1 - line.y0,
+                        }
+
+                        table.insert(rects, rect)
+                        line_last_rect = rect
+                    else
+                        if line_text_direction > 0 then -- left to right
+                            line_last_rect.w = word.x1 - line_last_rect.x
+                        else -- right to left
+                            line_last_rect.w = line_last_rect.w + (line_last_rect.x - word.x0)
+                            line_last_rect.x = word.x0
+                        end
+                    end
+
+                    if line_index == end_line_index and word_index == end_word_index then
+                        break
+                    end
+                end
+            end
+        end
+    end
+
+    if found_start then
+        return table.concat(words, " "), rects
+    else
+        return nil, nil
+    end
+end
+
+local function areTextBoxesEqual(boxes1, text1, boxes2, text2)
+    if text1 ~= text2 then
+        return false
+    end
+
+    if boxes1 and boxes2 then
+        if #boxes1 ~= #boxes2 then
+            return false
+        end
+
+        for i = 1, #boxes1, 1 do
+            if boxes1[i] ~= boxes2[i] then
+                return false
+            end
+        end
+
```

Comment:
Bunch of random blank lines around here and the previous local functions - and in HtmlBoxWidget:setPageNumber() and here and there. In a large functions or loops doing multiple things in steps, it's fine to separate the steps - in little ones, it's just useless and makes it bigger than it could be.
(And think about people with small computer screens reviewing :))

Reply:
Otherwise, fine with this PR.

Reply:
Okay, I removed even more blanks lines.